### PR TITLE
vscode-extensions.vadimcn.vscode-lldb: 1.5.3 -> 1.6.1 and refactor

### PR DESCRIPTION
**Pull request by oxalica**
###### Motivation for this change

1. vscode-lldb expects a special patched lldb to correctly print rust types. Upstream lldb cannot correctly handle rust enums.
2. Bump version to 1.6.1 and sync compile options.
3. Refactor and simplify.
4. Add semi-automatic update script, which doesn't update patched lldb. Manual check is still required.

Tested to debug rust programs with vscodium. Pretty printers of `Option` and `Vec` both work.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  - 352.8M (Patched lldb takes 344.5M)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+let
+  rev = "7f0cd38dae5cd77ac58adf97896ac0f41af759fc";
+  sha256 = "1z45wz0cfghirp2pwj6q8yqzpaafx5m7h206aj1wnqqb1k7qvwx4";
+
+  inherit (llvmPackages_8) llvm clang-unwrapped;
+
+in stdenv.mkDerivation {
+  name = "lldb-rust-enabled";
+  version = "git-${rev}";
```
Comment (midchildan) [inline pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix:11]: The version should be `unstable-2018-12-06` per the [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#sec-package-naming).

**Review by midchildan — CHANGES_REQUESTED** [review]

File: pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix (hunk 1 above)
Comment (midchildan) [reply pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix:11]: > If a package is not a release but a commit from a repository, then the version part of the name must be the date of that (fetched) commit. The date must be in "YYYY-MM-DD" format. Also append "unstable" to the name - e.g., "pkgname-unstable-2014-09-23".

**Review by midchildan — APPROVED** [review]

File: pkgs/misc/vscode-extensions/vscode-lldb/build-deps/package.json [hunk 2]
```diff
@@ -1,24 +1,25 @@
...
-		"typescript": "^3.7.0",
-		"vsce": "^1.73.0",
-		"vscode-debugadapter-testsupport": "^1.35.0",
-		"vscode-debugprotocol": "^1.35.0",
-		"webpack": "^4.39.1",
-		"webpack-cli": "^3.3.7",
-		"yauzl": "^2.10.0"
-	}
+  "name": "vscode-lldb",
+  "version": "1.5.3",
```
Comment (ggreif) [inline pkgs/misc/vscode-extensions/vscode-lldb/build-deps/package.json:3]: Shouldn't this be bumped too?

File: pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
+{ stdenv, fetchFromGitHub, llvmPackages_8, cmake, python3, swig, libedit, breakpointHook }:
+
+let
+  version = "2018-12-06";
+  rev = "7f0cd38dae5cd77ac58adf97896ac0f41af759fc";
+  sha256 = "1z45wz0cfghirp2pwj6q8yqzpaafx5m7h206aj1wnqqb1k7qvwx4";
+
+  inherit (llvmPackages_8) llvm clang-unwrapped;
```
Comment (ggreif) [inline pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix:8]: why not `llvmPackages_latest`?

Comment (oxalica) [reply pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix:8]: The forked lldb should be compiled with the same version llvm. There are lots of API change between different llvm version.
It's hard to patch the source for latest llvm. (Need to dump all commits between official lldb_8..lldb_latest)

Comment (oxalica) [reply pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix:10]: I added a comment about this.

https://github.com/NixOS/nixpkgs/blob/083c45e5c465b21d168eb58679c4f0bcc7cd7879/pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix#L8-L10

File: pkgs/misc/vscode-extensions/vscode-lldb/build-deps/package.json (hunk 2 above)
Comment (oxalica) [reply pkgs/misc/vscode-extensions/vscode-lldb/build-deps/package.json]: Fixed

File: pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix (hunk 3 above)
Comment (ggreif) [reply pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix:10]: Okay. If this builds and works (i.e. debugs well the Rust code) then let's get it merged soon. Then we can iterate.
I was able to swap in an `lldb_10` from the Rust repo (with modifications), so this shouldn't be too hard.

**Comment by oxalica** [comment]
ping @jonringer

File: pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix (hunk 3 above)
Comment (oxalica) [reply pkgs/misc/vscode-extensions/vscode-lldb/lldb/default.nix]: @ggreif It is refactored and now uses https://github.com/vadimcn/llvm-project/tree/codelldb/11.0.0 which is based on lldb_11

**Comment by jonringer** [comment]
oops sorry. Haven't had as much time for nix.

You have some conflicts :(

**Comment by jonringer** [comment]
also, ofborg is failing:
```
anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-4/pkgs/development/node-packages/node-env.nix:3:1 called without required argument 'utillinux', 
```

**Comment by oxalica** [comment]
@jonringer Rebased. 

> You have some conflicts :(

The nodePackages on master is updated quite frequently. I think it's better to at least have someone reviewed other parts before making tons of rebases.

**Comment by oxalica** [comment]
Revert changes about `utillinux` and `util-linux` according to  https://github.com/NixOS/nixpkgs/pull/108596/#pullrequestreview-562887453
Unfortunately these two PR conflicts with each other. :cry:  I'm not sure whether to merge these changes, they are not quite relevant.

**Comment by SuperSandro2000** [comment]
@oxalica please fix the merge conflict now after the other PR got merged.

**Comment by oxalica** [comment]
@SuperSandro2000 Rebased

**Comment by jonringer** [comment]
still conflicts... wonder why there's so many conflicts

**Comment by oxalica** [comment]
@jonringer Just due to update of adjacent vscode extensions. Rebased.
Another issue is that currently we prefer a mass single global nodePackages to reuse packages and reduce nixpkgs tree size, but it suffers from frequent merge conflicts and a total regenerate (mass diff) of the whole file in any PR touching it.
Related: #86776

**Comment by SuperSandro2000** [comment]
> frequent merge conflicts

That is an understatement. You get a merge conflict if anyone touches the file in the meanwhile.

File: pkgs/misc/vscode-extensions/vscode-lldb/default.nix [hunk 4]
```diff
@@ -1,28 +1,27 @@
...
 assert lib.versionAtLeast python3.version "3.5";
 let
   publisher = "vadimcn";
-  name = "vscode-lldb";
-  version = "1.5.3";
-
-  dylibExt = stdenv.hostPlatform.extensions.sharedLibrary;
+  pname = "vscode-lldb";
+  version = "1.6.1";
+  sha256 = "sha256-mi+AeHg9zO0vjF0OZCufPkliInqxTvDGV350wqAwe90=";
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/vscode-lldb/default.nix:9]:
```suggestion
```

File: pkgs/misc/vscode-extensions/vscode-lldb/default.nix [hunk 5]
```diff
@@ -1,28 +1,27 @@
...
+  cargoSha256 = "sha256-HPVbqYsst/iFrHn5wvmWtqeVHOHR7JT8lu+/xZq1lK0=";
 
   src = fetchFromGitHub {
     owner = "vadimcn";
     repo = "vscode-lldb";
     rev = "v${version}";
-    sha256 = "1139945j3z0fxc3nlyvd81k0ypymqsj051idrbgbibwshpi86y93";
+    inherit sha256;
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/vscode-lldb/default.nix:16]:
```suggestion
    sha256 = "sha256-mi+AeHg9zO0vjF0OZCufPkliInqxTvDGV350wqAwe90=";
```

File: pkgs/misc/vscode-extensions/vscode-lldb/default.nix (hunk 4 above)
Comment (oxalica) [reply pkgs/misc/vscode-extensions/vscode-lldb/default.nix]: Inlined.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101179` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
22 packages built:
</summary>
<ul>
<li>epgstation</li>
<li>flood</li>
<li>gtop</li>
<li>joplin</li>
<li>mirakurun</li>
<li>netlify-cli</li>
<li>pm2</li>
<li>postcss-cli</li>
<li>redoc-cli</li>
<li>vimPlugins.coc-diagnostic</li>
<li>vimPlugins.coc-eslint</li>
<li>vimPlugins.coc-git</li>
<li>vimPlugins.coc-go</li>
<li>vimPlugins.coc-html</li>
<li>vimPlugins.coc-metals</li>
<li>vimPlugins.coc-prettier</li>
<li>vimPlugins.coc-rust-analyzer</li>
<li>vimPlugins.coc-solargraph</li>
<li>vimPlugins.coc-stylelint</li>
<li>vimPlugins.coc-tsserver</li>
<li>vscode-extensions.matklad.rust-analyzer</li>
<li>vscode-extensions.vadimcn.vscode-lldb</li>
</ul>
</details>

**Comment by oxalica** [comment]
Rebased. ping @jonringer

**Comment by SuperSandro2000** [comment]
@oxalica you did not resolve all merge conflicts.

**Comment by oxalica** [comment]
@SuperSandro2000 The new conflict is caused by #110813. Rebased again now.

File: pkgs/misc/vscode-extensions/vscode-lldb/update.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p git jq nix-prefetch
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/vscode-lldb/update.sh:2]:
```suggestion
#!nix-shell -i bash -p jq nix-prefetch
```

File: pkgs/misc/vscode-extensions/default.nix [hunk 7]
```diff
@@ -689,9 +689,7 @@ let
         };
       };
 
-      vadimcn.vscode-lldb = callPackage ./vscode-lldb {
-        lldb = llvmPackages_latest.lldb;
-      };
+      vadimcn.vscode-lldb = callPackage ./vscode-lldb {};
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/default.nix:692]:
```suggestion
      vadimcn.vscode-lldb = callPackage ./vscode-lldb { };
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101179` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package marked as broken and skipped:
</summary>
<ul>
<li>image_optim</li>
</ul>
</details>
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
28 packages built:
</summary>
<ul>
<li>antora</li>
<li>bitwarden-cli</li>
<li>castnow</li>
<li>create-cycle-app</li>
<li>emojione</li>
<li>epgstation</li>
<li>etcher</li>
<li>flood</li>
<li>joplin</li>
<li>mirakurun</li>
<li>netlify-cli</li>
<li>parity-ui</li>
<li>postcss-cli</li>
<li>pulp</li>
<li>redoc-cli</li>
<li>ripcord</li>
<li>slack</li>
<li>teleprompter</li>
<li>thelounge</li>
<li>twemoji-color-font</li>
<li>vimPlugins.coc-metals</li>
<li>vimPlugins.coc-prettier</li>
<li>vimPlugins.coc-stylelint</li>
<li>vimPlugins.coc-tslint</li>
<li>vimPlugins.coc-vetur</li>
<li>vscode-extensions.matklad.rust-analyzer</li>
<li>vscode-extensions.vadimcn.vscode-lldb</li>
<li>wasm-text-gen</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101179` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>vscode-extensions.vadimcn.vscode-lldb: log was empty</li>
  </ul>
</details>
<details>
<summary>
1 package built:
</summary>
<ul>
<li>vscode-extensions.matklad.rust-analyzer</li>
</ul>
</details>

```
error: --- Error -----------------------------------------------------------------------------------------------------------------------------------------------------------------------
nixbuilder for '/nix/store/armn7hc6ljgm5m1k33irk00qnlgbsrk1-vscode-lldb-adapter-1.6.1.drv' failed with exit code 101; last 10 log lines:
___cpp_closure_17014716006298275771 in liblldb-b0d0b3ec9b2efe72.rlib(cpp_closures.o)
ld: symbol(s) not found for architecture x86_64
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
error: aborting due to previous error
error: could not compile `codelldb`
To learn more, run the command again with --verbose.
error: --- Error -----------------------------------------------------------------------------------------------------------------------------------------------------------------------
nix1 dependencies of derivation '/nix/store/755mzzqcahpph93xkqqi5zpvd9hdix04-vscode-extension-vscode-lldb.drv' failed to build
error: --- Error -----------------------------------------------------------------------------------------------------------------------------------------------------------------------
nix1 dependencies of derivation '/nix/store/h7k7p26l4mkki727d5ql76draffhyfj3-env.drv' failed to build
```

**Comment by oxalica** [comment]
@SuperSandro2000 Sorry but I don't have any darwin machine for test. Currently I just limit supported platforms to x86_64-linux only.

File: pkgs/misc/vscode-extensions/vscode-lldb/default.nix [hunk 8]
```diff
@@ -1,106 +1,101 @@
...
+  };
 
   meta = with lib; {
     description = "A native debugger extension for VSCode based on LLDB";
     homepage = "https://github.com/vadimcn/vscode-lldb";
     license = with licenses; [ mit ];
     maintainers = with maintainers; [ oxalica ];
-    platforms = platforms.all;
+    # Build failed on x86_64-darwin.
+    platforms = [ "x86_64-linux" ];
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/vscode-lldb/default.nix:99]: Build failures should be represented by broken and platforms should be what upstream theoretically supports. There is a darwin release upstream so this should be broken https://github.com/vadimcn/vscode-lldb/releases.

```suggestion
    platforms = platforms.all;
    broken = stdenv.isDarwin;
```

Comment (oxalica) [reply pkgs/misc/vscode-extensions/vscode-lldb/default.nix]: @SuperSandro2000 Got it. Applied.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101179` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package marked as broken and skipped:
</summary>
<ul>
<li>image_optim</li>
</ul>
</details>
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
28 packages built:
</summary>
<ul>
<li>antora</li>
<li>bitwarden-cli</li>
<li>castnow</li>
<li>create-cycle-app</li>
<li>emojione</li>
<li>epgstation</li>
<li>etcher</li>
<li>flood</li>
<li>joplin</li>
<li>mirakurun</li>
<li>netlify-cli</li>
<li>parity-ui</li>
<li>postcss-cli</li>
<li>pulp</li>
<li>redoc-cli</li>
<li>ripcord</li>
<li>slack</li>
<li>teleprompter</li>
<li>thelounge</li>
<li>twemoji-color-font</li>
<li>vimPlugins.coc-metals</li>
<li>vimPlugins.coc-prettier</li>
<li>vimPlugins.coc-stylelint</li>
<li>vimPlugins.coc-tslint</li>
<li>vimPlugins.coc-vetur</li>
<li>vscode-extensions.matklad.rust-analyzer</li>
<li>vscode-extensions.vadimcn.vscode-lldb</li>
<li>wasm-text-gen</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101179` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package marked as broken and skipped:
</summary>
<ul>
<li>image_optim</li>
</ul>
</details>
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
28 packages built:
</summary>
<ul>
<li>antora</li>
<li>bitwarden-cli</li>
<li>castnow</li>
<li>create-cycle-app</li>
<li>emojione</li>
<li>epgstation</li>
<li>etcher</li>
<li>flood</li>
<li>joplin</li>
<li>mirakurun</li>
<li>netlify-cli</li>
<li>parity-ui</li>
<li>postcss-cli</li>
<li>pulp</li>
<li>redoc-cli</li>
<li>ripcord</li>
<li>slack</li>
<li>teleprompter</li>
<li>thelounge</li>
<li>twemoji-color-font</li>
<li>vimPlugins.coc-metals</li>
<li>vimPlugins.coc-prettier</li>
<li>vimPlugins.coc-stylelint</li>
<li>vimPlugins.coc-tslint</li>
<li>vimPlugins.coc-vetur</li>
<li>vscode-extensions.matklad.rust-analyzer</li>
<li>vscode-extensions.vadimcn.vscode-lldb</li>
<li>wasm-text-gen</li>
</ul>
</details>

**Pull request merged**